assorted minors (util mailer func, datetime parse robust, test fix)

JohnTheodore opened the pull request:
I'm not sure if I've overlooked something here? it seems like we can have half as much stuff sent to terminal output while still giving the viewer all the same information?

JohnTheodore commented:
Before:
<img width="1119" alt="screen shot 2017-05-04 at 1 47 18 am" src="https://cloud.githubusercontent.com/assets/1904409/25696364/ca1ae0f6-306b-11e7-920e-d8e2b77c2650.png">

JohnTheodore commented:
After this PR:
<img width="1305" alt="screen shot 2017-05-04 at 1 43 54 am" src="https://cloud.githubusercontent.com/assets/1904409/25696391/e0d7b04e-306b-11e7-84ea-d4011821ad32.png">

JohnTheodore commented on c7n/policy.py:
```diff
@@ -201,24 +201,26 @@ def run(self, *args, **kw):
             return
 
         with self.policy.ctx:
+            region        = (self.policy.options.region or 'default').ljust(14, ' ')
```
the longest aws region string is 14 characters (any of the asia pacific regions)

kapilt commented:
its a before and after the thing when dealing with a policy thats taking a long time. i.e.. say something hangs, what was it... without a before there's no clear Indication in the logs.

JohnTheodore commented on c7n/policy.py:
```diff
@@ -201,24 +201,26 @@ def run(self, *args, **kw):
 
         with self.policy.ctx:
+            region        = (self.policy.options.region or 'default').ljust(14, ' ')
+            resource      = self.policy.resource_type.ljust(14, ' ')
```
the longest c7n resource string is elasticsearch, 14 characters also

kapilt commented:
that said I kinda of dig the output screenshots, whats the width on these rows

JohnTheodore commented on c7n/policy.py:
```diff
@@ -201,24 +201,25 @@ def run(self, *args, **kw):
+                    resource,
+                    qty_resources,
+                    run_time,
+                    self.policy.name
```
self.policy.name is last, since it's length is arbitrary (sometimes quite long, like 50 chars). so if we keep this last, all the other shell output columns will align in a neatly human readable way.

kapilt replied:
i like the region inclusion, but the shifting the policy to the right edge when its probably the most important aspect of the info seems unfortunate.

kapilt commented:
that sounds good re log level.debug on the before run

JohnTheodore commented:
The longest row width in the before shot is 158 chars
The longest row width in the    after shot is 168 chars

I happen to name my policies somewhat long names, and I still have a bunch of room in iterm for space to the right.

JohnTheodore commented:
ok, I put the before log back how it was, changed it from info to debug.
I just tested this, it all prints neatly and is easy to read.

I squashed all this into 1 commit, and I'm passing 3/3 checks.

JohnTheodore commented on tests/test_logsub.py:
```diff
@@ -32,7 +32,8 @@ def setUp(self):
 
     def tearDown(self):
-        os.remove('config.json')
+        if os.path.isfile('config.json'):
```
I think this fixes the occasional/random test failure that comes from test_logsub.py

JohnTheodore commented on c7n/filters/core.py:
```diff
@@ -432,8 +432,7 @@ def process_value_type(self, sentinel, value):
                     value = parse(value, default=datetime.now(tz=tzutc()))
-
-                except (AttributeError, TypeError):
+                except:
```
dateutil.parse will raise exceptions like ValueError and other things commonly. It's probably better if we catch all exceptions instead of white listing two kinds of errors. I noticed one of my users put garbage in an 'ExpiresAt' tag that gets sent to this function.

kapilt replied:
i'd rather just add the ValueError to the list, afaics between the three that should cover. unconditional exceptions can mask other things. at the very least on a blank it needs an inline comment.

JohnTheodore replied:
fixed

JohnTheodore commented on tools/c7n_mailer/README.md:
```diff
@@ -248,7 +248,8 @@ The following extra jinja filters are available:
 | filter | behavior |
 |:----------|:-----------|
-| `{{ utc_string|date_time_format(tz_str='US/Pacific', format='%Y %b %d %H:%M %Z') }}` | pretty [format](https://docs.python.org/2/library/datetime.html#strftime-strptime-behavior) the date / time |
+| <code>utc_string&#124;date_time_format(tz_str='US/Pacific', format='%Y %b %d %H:%M %Z')</code> | pretty [format](https://docs.python.org/2/library/datetime.html#strftime-strptime-behavior) the date / time |
```
This fixes a rendering issue with markdown on the README. I didn't realize I couldn't use a real pipe, since that messed up how it's viewed. I switched to using the html code for a pipe: &#124;

JohnTheodore commented on tools/c7n_mailer/c7n_mailer/record.py:
```diff
@@ -21,6 +24,10 @@ def date_time_format(utc_str, tz_str='US/Eastern', format='%Y %b %d %H:%M %Z'):
     return parser.parse(utc_str).astimezone(gettz(tz_str)).strftime(format)
 
 
+def get_date_time_delta(delta):
```
`
{% macro pretty_print_datetime(datetime_string, resource, action, policy, region) -%}
  {{ datetime_string|date_time_format(tz_str='US/Pacific', format='%Y %b %d %H:%M %Z') }}
{%- endmacro %}

{% macro get_maid_future_action(column_name, resource, action, policy, region) -%}
  {% for action in policy['actions'] %}
    {% if action['type'] == 'mark-for-op' %}
      {% set future_action_date = action['days']|get_date_time_delta() %}
      {{ action['op'] }}@{{ pretty_print_datetime(future_action_date) }}
    {% endif %}
  {% endfor %}
{%- endmacro %}
`

So now we can render any of the UTC strings from AWS's apis (creation dates, etc), and also render policy action day deltas, as a pretty date.

JohnTheodore commented on c7n/policy.py:
```diff
@@ -80,7 +80,7 @@ def __init__(self, data, options):
                 options.regions = all_regions
             for region in options.regions:
-                if region not in all_regions:
+                if len(all_regions) > 0 and region not in all_regions:
```
This is a duct tape patch to get c7n working again. It should be removed when a better fix is put in place.

kapilt replied:
working on that now in #1137

JohnTheodore replied:
I reverted this duct tape for this PR. Let me know if there are any changes you want in here.

JohnTheodore commented on tools/c7n_mailer/c7n_mailer/sqs_message_processor.py:
```diff
@@ -260,9 +268,9 @@ def render(self, target, data, resources):
-                data['action'].get('template', 'default')))
+        except Exception as error_msg:
+            self.logger.error("Invalid template reference %s.j2\n%s" % (
+                data['action'].get('template', 'default')), error_msg)
```
by bubbling up the actual error from jinja, it's easier to find out what the problems are for people looking at lambda mailer logs.

kapilt commented on c7n/policy.py:
```diff
@@ -229,24 +229,31 @@ def run(self, *args, **kw):
-            s = time.time()
-            resources = self.policy.resource_manager.resources()
-            rt = time.time() - s
+            region        = (self.policy.options.region or 'default').ljust(14, ' ')
```
i'm a little wary of the hard coded alignments here, and also of mixing more presentation logic into the logs. 

afaics the root of the value chain here is just switching the info to debug on the double print.

JohnTheodore replied:
nod, understood. I think it's aesthetic preference. I can remove the ljust and rjusts.

I will say, the one biggest benefit I've had.. is when I'm parsing logs as part of testing and other situations, having that count column in the same place always is super helpful. If I'm running a test of 100 policies, it's pretty easy to scan the terminal output and see that count.

Between count, time, policy, region and resource what order do you generally prefer?

If we remove all the ljust/rjusts except for count, and put count in the first column.. it would still be easy to find which policies have count easily when scanning a large list. (This could work with no ljust or just, but having 1 ljust or just on count for say ljust(4, ' ') could be nice, or 3. Thoughts?

JohnTheodore replied:
Do you guys have UI/UX people that can advise us on the intricacies of terminal output formatting? :p

kapilt commented on c7n/policy.py:
```diff
@@ -229,24 +229,31 @@ def run(self, *args, **kw):
-            rt = time.time() - s
+            region        = (self.policy.options.region or 'default').ljust(14, ' ')
+            resource      = self.policy.resource_type.ljust(14, ' ')
+            start         = time.time()
```
what's up this source margin / justify on equality? the extra whitespace here isn't per the project style.

JohnTheodore commented on c7n/filters/core.py:
```diff
@@ -459,7 +458,10 @@ def process_value_type(self, sentinel, value):
 
             if not isinstance(value, datetime):
-                value = parse(value, default=datetime.now(tz=tzutc()))
+                try:
```
I normalized this one to look the same as the one above, and also filter for the 3 error types.

JohnTheodore commented on c7n/policy.py:
```diff
@@ -244,24 +244,31 @@ def run(self, *args, **kw):
+            qty_resources = str(len(resources))
             self.policy.log.info(
-                "policy: %s resource:%s has count:%d time:%0.2f" % (
+                "count: %s for policy: %s resource: %s region: %s time: %0.2f" % (
```
I removed any ljust or rjust stuff from this section. let me know if there are any other changes you want on this PR

kapilt commented on c7n/policy.py:
```diff
@@ -257,19 +257,25 @@ def run(self, *args, **kw):
-            s = time.time()
-            resources = self.policy.resource_manager.resources()
-            rt = time.time() - s
+            region        = self.policy.options.region or 'default'
```
could you remove the column alignment

JohnTheodore replied:
fixed

JohnTheodore commented:
Let me know if you want other changes.

kapilt commented on c7n/policy.py:
```diff
@@ -257,19 +257,25 @@ def run(self, *args, **kw):
             resources = self.policy.resource_manager.resources()
-            rt = time.time() - s
+            run_time = time.time() - start
+            qty_resources = str(len(resources))
```
can we just remove this log change, its got a lot of extraneous transient variables, and trunk has changed to include region, afaics the only delta is spacing formatting on the log string, the information conveyed is the same.

JohnTheodore replied:
ahh, you pre-empted the log.debug with your own. and with removing the ljust and rjusts, this whole file should just go back to the way HEAD is and be untouched. It made more sense in the original before the changes to the branch and HEAD.

JohnTheodore replied:
fixed

JohnTheodore commented:
I rebased from master and squashed this into 1 commit.

JohnTheodore commented on c7n/filters/core.py:
```diff
@@ -434,8 +434,7 @@ def process_value_type(self, sentinel, value):
                     value = parse(value, default=datetime.now(tz=tzutc()))
-
-                except (AttributeError, TypeError):
+                except (AttributeError, TypeError, ValueError):
```
File "/Users/john_theodore/code/cloud-custodian/lib/python2.7/site-packages/dateutil/parser.py", line 559, in parse
    raise ValueError("Unknown string format")
ValueError: Unknown string format

I'm still getting these. :(

kapilt replied:
from this call site? ie, what's next up the stack
